### DISCUSSION TITLE
Never ending loading on iOS

### DISCUSSION THREAD

**Discussion by HBubli** [PrivateBin/PrivateBin#1405] [Q&A] [answered]
### Describe the problem/question

My privatebin instance works fine everywhere except on my iPad where it doesnt work in any browser, even with cache cleared and everything disabled.

### Did you use the FAQ section?

- [X] Yes, I have read [the FAQ](https://github.com/PrivateBin/PrivateBin/wiki/FAQ) and I found no solution/answer there.

### What you did?

1. Opened the instance in safati
2. Wait until bored
3. Doesnt load

### What happens

The “Loading… In case this message never disappears please have a look at [this FAQ for information to troubleshoot.” Message is displayed and doesn’t disappear

### What should happen

Privatebin loads up normally

### Additional information

I have also tried reinstalling privatebin several times.

### Server address

https://bin.hbubli.cc

### Server OS

Ubuntu 22.04

### Webserver

Apache + Nginx as proxy

### PrivateBin version

1.7.4

### Browser and version

Safari 17.6.1

### Local operating system and version

iPadOS 17.6.1

### Issue reproducibility

No, I cannot reproduce it on https://privatebin.net.

**Comment by rugk** [comment]
> No, I cannot reproduce it on https://privatebin.net/.

If that is the case, it obviously should be by some configuration problem or other problem on your instance. Though that it works on all other non Safari browsers (AFAIK no browser on iOS uses [non-Safari browser rendering engine yet](https://www.theverge.com/2024/1/25/24050478/apple-ios-17-4-browser-engines-eu)) is indeed strange.

Do you by change have a desktop Safari or can debug the mobile browser to e.g. see console errors?

Though, okay, I have GNOME Web (epiphany; which also uses WebKit aka the same browser engine as Safari) installed and I could reproduce the issue there:
![grafik](https://github.com/user-attachments/assets/0e3e040c-d037-451d-8abd-fcf0d662c469)

Console errors seem to indicate something with your jQuery version is wrong(?) or jQuery is not loaded correctly?:
```
[Error] Unrecognized Content-Security-Policy directive 'manifest-src'.

[Error] Unrecognized Content-Security-Policy directive 'manifest-src'.

[Error] Error: Bootstrap's JavaScript requires jQuery
	Global Code (bootstrap-3.4.1.js:6:89)
[Error] ReferenceError: Can't find variable: jQuery
	Global Code (privatebin.js:15)
``` 

I can confirm, it works in Firefox.

If everything fails, the PrivateBin devs also have access to browserstacks for testing different devices, so just in case.

**Reply by elrido** [reply]
I could reproduce it on browserstack in safari on MacOS on that instance as well, console shows the same thing - weirdly it works fine for privatebin.net.

AFAIK this means that this browser engine doesn't respect the async and defer flags. We didn't use them as broadly until #1372 were I added them to all scripts. That change has not yet made it into a release.

Safari and iOS should have support for them, according to https://caniuse.com/script-defer - Are we not using them correctly?

**Reply by HBubli** [reply]
> I could reproduce it on browserstack in safari on MacOS on that instance as well, console shows the same thing - weirdly it works fine for privatebin.net.
> 
> AFAIK this means that this browser engine doesn't respect the async and defer flags. We didn't use them as broadly until #1372 were I added them to all scripts. That change has not yet made it into a release.
> 
> Safari and iOS should have support for them, according to https://caniuse.com/script-defer - Are we not using them correctly?

Is there something i can do to fix the issue on my instance?

**Reply by elrido** [reply]
> Is there something i can do to fix the issue on my instance?

For now, I would recommend to either downgrade to a stable release (1.7.4 doesn't yet seem to be affected). What you or one of us should do to isolate what causes it is to experiment under Safari with removing the async and defer attributes from the script-lines in the template used. Instead of `<?php $this->_scriptTag('js/privatebin.js', 'defer'); ?>` it would become `<?php $this->_scriptTag('js/privatebin.js'); ?>` or `<?php $this->_scriptTag('js/privatebin.js', ''); ?>`. My current assumption is that the Safari engine for some reason doesn't `defer` loading of privatebin and bootstrap libraries and they get launched before jquery is loaded, leading to the error that you see in the Safari console. I haven't had the time, yet, to perform that experiment myself.

**Reply by HBubli** [reply]
1.7.4 seems to work fine on my iPad, ima stay on that version for now. Thanks for your help.

**Reply by rugk** [reply]
Unfortunately, that does not solve the issue for any other users.

Still I don't get why it works on PrivateBin.net then. Should not the issue occur there, too?

**Reply by rugk** [reply]
Ah ookay, because of the versions…

Trying a public codespace for now, I can reproduce it: https://obscure-parakeet-g44wvxv9pww2wwp5-8080.app.github.dev/
702831ea38199ad19884b758faade176bae1f959

**Reply by rugk** [reply, marked as answer]
As this is certainly a (big) problem I created an issue in https://github.com/PrivateBin/PrivateBin/issues/1423 and will continue there…
